src/webgpu/util/binary_stream.ts

```diff
@@ -0,0 +1,283 @@
+import { assert } from '../../common/util/util.js';
+import { Float16Array } from '../../external/petamoriken/float16/float16.js';
+
+import { align } from './math.js';
+
+/**
+ * BinaryStream is a utility to efficiently encode and decode numbers to / from a Uint8Array.
+ * BinaryStream uses a number of internal typed arrays to avoid small array allocations when reading
+ * and writing.
+ */
+export default class BinaryStream {
+  /**
+   * Constructor
+   * @param buffer the buffer to read from / write to. Array length must be a multiple of 8 bytes.
+   */
+  constructor(buffer: Uint8Array) {
+    this.offset = 0;
+    this.u8 = buffer;
+    this.u16 = new Uint16Array(this.u8.buffer);
+    this.u32 = new Uint32Array(this.u8.buffer);
+    this.i8 = new Int8Array(this.u8.buffer);
+    this.i16 = new Int16Array(this.u8.buffer);
+    this.i32 = new Int32Array(this.u8.buffer);
+    this.f16 = new Float16Array(this.u8.buffer);
+    this.f32 = new Float32Array(this.u8.buffer);
+    this.f64 = new Float64Array(this.u8.buffer);
```

Comment:
would it be simpler with DataView?

blog says it should be just as fast as native typed arrays if you use little-endian
https://v8.dev/blog/dataview

Reply:
Maybe. I briefly looked at `DataView` but reached for what I knew worked. I don't mind trying `DataView` as a followup change, to see if there are any perf wins.